### PR TITLE
feat: lint dockerfiles

### PR DESCRIPTION
**Pull request by goller**
Many times it is difficult to lint dockerfiles during a build or bake.
This adds a new CLI option `--lint` to build and bake that runs hadolint
on any dockerfiles.

`--lint` by default will fail the build on any lints that are considered errors.

The lint failure level can be controlled with the `--lint-on-fail` flag with these options:

* `error`: fail the build if any lints are error level.
* `warn`: fail if any lints are warning or error.
* `info`: fail if any lints are info, warning or error.
* `none`: do not fail the build for linter issues.

### CODE REVIEW

File: pkg/buildx/commands/bake.go [hunk 1]
```diff
@@ -115,14 +115,30 @@ func RunBake(dockerCli command.Cli, in BakeOptions, validator BakeValidator) (er
 	}
 
 	buildxNodes := builder.ToBuildxNodes(nodes)
+	buildxNodes, err = build.FilterAvailableNodes(buildxNodes)
```
Comment (goller) [inline pkg/buildx/commands/bake.go:118]: We now start boot the drivers _before_ we bake or build.

File: pkg/buildx/commands/bake.go [hunk 2]
```diff
@@ -161,13 +177,14 @@ func RunBake(dockerCli command.Cli, in BakeOptions, validator BakeValidator) (er
 			if in.exportLoad {
 				progress.Write(printer, "[load] fast load failed; retrying", func() error { return err })
 				buildOpts, _ = load.WithDepotImagePull(fallbackOpts, load.DepotLoadOptions{})
-				_, err = build.DepotBuild(ctx, buildxNodes, buildOpts, dockerClient, dockerConfigDir, printer)
+				_, err = build.DepotBuild(ctx, buildxNodes, buildOpts, dockerClient, dockerConfigDir, printer, nil)
```
Comment (goller) [inline pkg/buildx/commands/bake.go:180]: No linting when we retry.

**Review by jacobwgillespie — COMMENTED** [review]
This is super awesome! 🎉 A few minor things:

- [x] I think we should have completions for the values of `--lint` (I pushed a commit that should add that)
- [ ] The "linter issues found" summary does not show whether the issues are info/warning/error like they do in the logs, I think we should include the severity of issues in the summary as well
- [ ] I wonder if we should separate out whether linting is enabled and which severity fails the build. Like maybe `--lint` should be a boolean flag, and if enabled linting happens and the build fails if any ERROR issues are found. I imagine that's the most common desired behavior. Then there could be some other flag that controls which severity fails the build, not sure what it should be called, maybe there's some prior art, but something like `--lint-fail-on warning`
- [ ] The `More info: URL` could use the new link thing from build summary to be clickable

**Comment by jacobwgillespie** [comment]
One other thought: right now, linter failures report as build failures in the API. I think conceptually that makes sense, but we should probably report whether `--lint` was used, with what options, in the `initiator` details, and should track whether the failures was in the lint or build phases, so that we can display that in the UI.

**Comment by goller** [comment]
> * [x]  The `More info: URL` could use the new link thing from build summary to be clickable

image

Ok! Added in [9ad2d60](https://github.com/depot/cli/pull/126/commits/9ad2d60f89ffd641a045ca0225524aa0d03bbb37)

**Comment by goller** [comment]
> * [x]  I wonder if we should separate out whether linting is enabled and which severity fails the build. Like maybe `--lint` should be a boolean flag, and if enabled linting happens and the build fails if any ERROR issues are found. I imagine that's the most common desired behavior. Then there could be some other flag that controls which severity fails the build, not sure what it should be called, maybe there's some prior art, but something like `--lint-fail-on warning`

Ok! `--lint` is now a bool and `--lint-fail-on` can be `error` (default), `warn`, `info`  or `none`.
[1f39f39](https://github.com/depot/cli/pull/126/commits/1f39f39e9decdc12885030ce206bdc01e42e7c71)

Here is an  example failing on info.

image

**Comment by goller** [comment]
> One other thought: right now, linter failures report as build failures in the API. I think conceptually that makes sense, but we should probably report whether `--lint` was used, with what options, in the `initiator` details, and should track whether the failures was in the lint or build phases, so that we can display that in the UI.

Added in [8895f2c](https://github.com/depot/cli/pull/126/commits/8895f2ca43e1f0e44704c40418f76f2ca30bf775)

**Pull request merged**